lib/web_ui/lib/src/engine/canvaskit/font_fallbacks.dart

```diff
@@ -5,65 +5,125 @@
+
+  /// Index of all font families by code unit range.
+  final IntervalTree<NotoFont> notoTree = createNotoFontTree();
+
+  static IntervalTree<NotoFont> createNotoFontTree() {
+    Map<NotoFont, List<CodeunitRange>> ranges =
+        <NotoFont, List<CodeunitRange>>{};
+
+    for (NotoFont font in _notoFonts) {
+      // TODO(yjbanov): instead of mutating the font tree during reset, it's
+      //                better to construct an immutable tree of resolved fonts
+      //                pointing back to the original NotoFont objects. Then
+      //                resetting the tree would be a matter of reconstrucint
+      //                the new resolved tree.
+      font.reset();
+      for (CodeunitRange range in font.approximateUnicodeRanges) {
+        ranges.putIfAbsent(font, () => <CodeunitRange>[]).add(range);
+      }
+    }
+
+    return IntervalTree<NotoFont>.createFromRanges(ranges);
+  }
 
-Future<void> findFontsForMissingCodeunits(List<int> codeunits) async {
-  _ensureNotoFontTreeCreated();
+
+  /// Fallback fonts which have been registered and loaded.
+  final List<_RegisteredFont> registeredFallbackFonts = <_RegisteredFont>[];
+
+  final List<String> globalFontFallbacks = <String>['Roboto'];
+
+  final Map<String, int> fontFallbackCounts = <String, int>{};
+
+  void registerFallbackFont(String family, Uint8List bytes) {
+    fontFallbackCounts.putIfAbsent(family, () => 0);
+    int fontFallbackTag = fontFallbackCounts[family]!;
+    fontFallbackCounts[family] = fontFallbackCounts[family]! + 1;
+    String countedFamily = '$family $fontFallbackTag';
+    registeredFallbackFonts.add(_RegisteredFont(bytes, countedFamily));
+    globalFontFallbacks.add(countedFamily);
+  }
+}
+
+Future<void> findFontsForMissingCodeunits(List<int> codeUnits) async {
+  final FontFallbackData data = FontFallbackData.instance;
 
   // If all of the code units are known to have no Noto Font which covers them,
   // then just give up. We have already logged a warning.
-  if (codeunits.every((u) => codeUnitsWithNoKnownFont.contains(u))) {
+  if (codeUnits.every((u) => data.codeUnitsWithNoKnownFont.contains(u))) {
     return;
   }
   Set<NotoFont> fonts = <NotoFont>{};
   Set<int> coveredCodeUnits = <int>{};
   Set<int> missingCodeUnits = <int>{};
-  for (int codeunit in codeunits) {
-    List<NotoFont> fontsForUnit = _notoTree!.intersections(codeunit);
+  for (int codeUnit in codeUnits) {
+    List<NotoFont> fontsForUnit = data.notoTree.intersections(codeUnit);
     fonts.addAll(fontsForUnit);
     if (fontsForUnit.isNotEmpty) {
-      coveredCodeUnits.add(codeunit);
+      coveredCodeUnits.add(codeUnit);
     } else {
-      missingCodeUnits.add(codeunit);
+      missingCodeUnits.add(codeUnit);
     }
   }
 
-  fonts = findMinimumFontsForCodeunits(coveredCodeUnits, fonts);
-
   for (NotoFont font in fonts) {
     await font.ensureResolved();
   }
 
+  final Set<int> unmatchedCodeUnits = Set<int>.from(coveredCodeUnits);
```

Comment:
this was confusing initially. maybe comment that it doesn't actually become "unmatchedCodeUnits" until it is passed as an input to `findMInimumFontsForCodeUnits`

Reply:
Done.